pilot/pkg/proxy/envoy/v2/ads_test.go

```diff
@@ -269,16 +284,112 @@ func TestAdsEds(t *testing.T) {
 	strResponse, _ := model.ToJSONWithIndent(res1, " ")
 	_ = ioutil.WriteFile(util.IstioOut+"/edsv2_sidecar.json", []byte(strResponse), 0644)
 
-	_ = server.EnvoyXdsServer.MemRegistry.AddEndpoint("service3.default.svc.cluster.local",
+	_ = server.EnvoyXdsServer.MemRegistry.AddEndpoint("adsupdate.default.svc.cluster.local",
 		"http-main", 2080, "10.1.7.1", 1080)
-	v2.PushAll() // will trigger recompute and push
+
+	// will trigger recompute and push for all clients - including some that may be closing
+	// This reproduced the 'push on closed connection' bug.
+	v2.PushAll()
 
 	res1, err = adsReceive(edsstr, 5*time.Second)
 	if err != nil {
 		t.Fatal("Recv2 failed", err)
 	}
 	strResponse, _ = model.ToJSONWithIndent(res1, " ")
 	_ = ioutil.WriteFile(util.IstioOut+"/edsv2_update.json", []byte(strResponse), 0644)
-
 	_ = edsstr.CloseSend()
 }
+
+// Make a direct EDS grpc request to pilot, verify the result is as expected.
+func TestAdsEds(t *testing.T) {
+	server := initLocalPilotTestEnv(t)
+	testAdsMultiple(t, server)
+}
+
+// Extract cluster load assignment from a discovery response.
+func getLoadAssignment(res1 *xdsapi.DiscoveryResponse) (*xdsapi.ClusterLoadAssignment, error) {
+	if res1.TypeUrl != "type.googleapis.com/envoy.api.v2.ClusterLoadAssignment" {
+		return nil, errors.New("Invalid typeURL" + res1.TypeUrl)
+	}
+	if res1.Resources[0].TypeUrl != "type.googleapis.com/envoy.api.v2.ClusterLoadAssignment" {
+		return nil, errors.New("Invalid resource typeURL" + res1.Resources[0].TypeUrl)
+	}
+	cla := &xdsapi.ClusterLoadAssignment{}
+	err := cla.Unmarshal(res1.Resources[0].Value)
+	if err != nil {
+		return nil, err
+	}
+	return cla, nil
+}
+
+func testIp(id uint32) string {
+	ipb := []byte{0, 0, 0, 0}
+	binary.BigEndian.PutUint32(ipb, id)
+	return net.IP(ipb).String()
+}
+
+func testAdsMultiple(t *testing.T, server *pilotbootstrap.Server) {
+	wg := &sync.WaitGroup{}
```

Comment:
Add `t.Helper()`

Reply:
Didn't know about t.Helper - but this is not a helper but a sub-test (or should be).

Reply:
It's not really a subtest as you have it structured here.  A subtest would be where you call `t.Run(testName, func(*testing.T)`, but that's not what is happening here.  In this case, it's a helper.